keras_rcnn/preprocessing/_object_detection.py

```diff
@@ -1,118 +1,84 @@
-import threading
-
+import keras.backend
+import keras.preprocessing.image
 import numpy
-import numpy.random
+import skimage.transform
 import skimage.io
-import keras.utils
-import sklearn.preprocessing
-
-
-class Iterator:
-    def __init__(self, n, batch_size, shuffle, seed):
-        self.batch_index = 0
-
-        self.batch_size = batch_size
-
-        self.index_generator = self._flow_index(n, batch_size, shuffle, seed)
-
-        self.lock = threading.Lock()
-
-        self.n = n
-
-        self.shuffle = shuffle
-
-        self.total_batches_seen = 0
-
-    def reset(self):
-        self.batch_index = 0
-
-    def _flow_index(self, n, batch_size=32, shuffle=False, seed=None):
-        self.reset()
-
-        while True:
-            if seed is not None:
-                numpy.random.seed(seed + self.total_batches_seen)
-
-            if self.batch_index == 0:
-                index_array = numpy.arange(n)
-
-                if shuffle:
-                    index_array = numpy.random.permutation(n)
+import time
 
-            current_index = (self.batch_index * batch_size) % n
+def scale_shape(shape, min_size=224, max_size=224):
+    """
+    Rescales a given shape such that the larger axis is no
+    larger than max_size and the smallest axis is as close
+    as possible to min_size.
+    """
+    min_shape = numpy.min(shape[0:2])
+    max_shape = numpy.max(shape[0:2])
+    scale     = min_size / min_shape
 
-            if n > current_index + batch_size:
-                current_batch_size = batch_size
+    # Prevent the biggest axis from being larger than max_size
+    if numpy.round(scale * max_shape) > max_size:
+        scale = max_size / max_shape
 
-                self.batch_index += 1
-            else:
-                current_batch_size = n - current_index
+    return (int(shape[0] * scale), int(shape[1] * scale), shape[2]), scale
 
-                self.batch_index = 0
 
-            self.total_batches_seen += 1
+def scale_image(image, min_size=224, max_size=224):
+    """
+    Rescales an image according to the heuristics from 'scale_shape'.
+    """
+    target_shape, scale = scale_shape(image.shape, min_size, max_size)
 
-            yield index_array[current_index:current_index + current_batch_size], current_index, current_batch_size
+    return skimage.transform.rescale(image, scale=scale, mode="reflect"), scale
 
-    def __iter__(self):
-        return self
 
-    def __next__(self, *args, **kwargs):
-        return self.next(*args, **kwargs)
+class DictionaryIterator(keras.preprocessing.image.Iterator):
+    def __init__(self, data, classes, image_data_generator, image_shape, batch_size=1,
```

Comment:
You don’t need to include the `image_shape` parameter since many datasets have variable shapes (e.g. PASCAL VOC) and this information is included in the dictionary.